docs/01-for-users: Add page about rebuilding a Flatpak from sources

travier opened the pull request:
Draft as this can probably be improved in some parts.

---

Rendered: https://github.com/travier/flathub-documentation/blob/main-rebuilding-flatpak/docs/01-for-users/10-rebuilding.md

bbhtt commented:
Docs is fine, I'm wondering whether this should be here or in https://docs.flatpak.org/en/latest/tips-and-tricks.html

travier commented:
I was wondering if this was Flathub specific or not. I don't know the answer to that question.

travier commented on docs/01-for-users/10-rebuilding.md:
````diff
@@ -0,0 +1,195 @@
+And then we can rebuild the Flatpak using:
+
+```bash
+flatpak-builder --disable-download --repo=repo --force-clean --disable-rofiles-fuse app org.kde.minuet.json
````
I could not get `--extra-sources=SOURCE-DIR` to work here, not sure why.

travier commented on docs/01-for-users/10-rebuilding.md:
````diff
@@ -0,0 +1,195 @@
+    └── org.kde.minuet.json
+```
+
+We can then copy the sources and patches from this folder into our current one:
````
See extra-sources comment below

bbhtt commented:
The sources extension isn't Flathub specific, and neither the fact that it is available. All someone needs to do is pass some args to ensure a `Sources` extension is produced and uploaded to the repo.

But some parts of the docs here are specific to Flathub since you are using an example from here.

bbhtt commented on docs/01-for-users/10-rebuilding.md:
````diff
@@ -0,0 +1,195 @@
+}
+```
+
+Notice that this manifests includes the exact commit of the Runtime it was
````
```suggestion
Notice that this manifest includes the exact commit of the runtime it was
```

bbhtt commented on docs/01-for-users/10-rebuilding.md:
````diff
@@ -0,0 +1,195 @@
+above):
+
+```bash
+flatpak --user install org.kde.Platform/x86_64/6.9
````
```suggestion
flatpak --user install flathub org.kde.Platform//6.9
```

same for below.

bbhtt commented on docs/01-for-users/10-rebuilding.md:
````diff
@@ -0,0 +1,195 @@
+flatpak-builder --disable-download --repo=repo --force-clean --disable-rofiles-fuse app org.kde.minuet.json
+```
+
+Note that you can reproduce this process for any version of the Flatpak by
+looking at the commit log and downgrading it to a specific commit. See
````
as long as it is available in the repo, we don't store everything forever.

bbhtt commented on docs/01-for-users/10-rebuilding.md:
````diff
@@ -0,0 +1,195 @@
+Let's first install the latest version:
+
+```bash
+flatpak --user install org.kde.minuet
````
```suggestion
flatpak --user install flathub org.kde.minuet
```

keszybz commented on docs/01-for-users/10-rebuilding.md:
````diff
@@ -0,0 +1,195 @@
+You can then find them in the Flatpak runtime folder:
+
+```bash
+tree ~/.local/share/flatpak/runtime/org.kde.minuet.Sources/x86_64/stable/fdafb56a6e907f87a359c8a048471fc1747291bdcbe0e952ad2d0cf3d66fc0dc/files
````
Maybe use `console` and `$ tree`. This should then get highlighted nicely.

bbhtt replied:
> I wanted to show the commit explicitly but active works as well.

ok that's fine then.

travier commented:
I *think* I've addressed most comments.

I wanted to add a section about bit for bit reproducibility of the build but I'm afraid that it's a per-app discussion, even though Flatpak tries really hard to make the builds reproducible.

bbhtt commented on docs/01-for-users/10-rebuilding.md:
```diff
@@ -0,0 +1,224 @@
+flatpak --user install flathub org.kde.Platform//6.9
+flatpak --user update \
+      --commit=f930fae18cfc829f51db18b9324905a3bebee0ec7e9d4d62afbb17f696fb20d0 \
+      org.kde.Platform/x86_64/6.9
```
```suggestion
      org.kde.Platform//6.9
```

bbhtt commented on docs/01-for-users/10-rebuilding.md:
```diff
@@ -0,0 +1,224 @@
+flatpak --user install org.kde.Sdk/x86_64/6.9
+flatpak --user update \
+      --commit=3170c974605b5af73a78bef2ae022df9b8dd7496569928a3766f0706c6c6515d \
+      org.kde.Sdk/x86_64/6.9
```
```suggestion
      org.kde.Sdk//6.9
```

travier commented:
Thanks, updated.
